virtool/uploads/data.py

```diff
@@ -0,0 +1,203 @@
+from logging import getLogger
+from typing import List, Optional, Union
+
+from sqlalchemy.ext.asyncio import AsyncSession
+from sqlalchemy import select, update
+from virtool_core.models.upload import UploadMinimal, UploadSearchResult
+
+import virtool.utils
+from virtool.data.errors import ResourceNotFoundError
+from virtool.data.piece import DataLayerPiece
+
+
+from virtool.mongo.core import DB
+from virtool.mongo.transforms import apply_transforms
+from virtool.uploads.db import finalize
+from virtool.uploads.models import Upload
+from virtool.users.db import AttachUserTransform
+from virtool.utils import run_in_thread, rm
+
+logger = getLogger(__name__)
+
+
+class UploadsData(DataLayerPiece):
+    def __init__(self, config, db, pg):
+        self._config = config
+        self._db: DB = db
+        self._pg = pg
+
+    async def find(
+        self, user: str = None, upload_type: str = None, ready: bool = None
+    ) -> UploadSearchResult:
+        """
+        Find and filter uploads.
+        """
+
+        filters = [Upload.removed == False]
+        uploads = []
+
+        async with AsyncSession(self._pg) as session:
+            if user:
+                filters.append(Upload.user == user)
+
+            if upload_type:
+                filters.append(Upload.type == upload_type)
+
+            if ready is not None:
+                filters.append(Upload.ready == ready)
+
+            results = await session.execute(select(Upload).filter(*filters))
+
+        for result in results.unique().scalars().all():
+            uploads.append(result.to_dict())
+
+        uploads = await apply_transforms(uploads, [AttachUserTransform(self._db)])
+
+        return UploadSearchResult(documents=uploads)
+
+    async def create(
+        self,
+        name: str,
+        upload_type: str,
+        reserved: bool = False,
+        user: Optional[str] = None,
+    ) -> Upload:
+        """
+        Create an upload.
+        """
+
+        async with AsyncSession(self._pg) as session:
+            upload = Upload(
+                created_at=virtool.utils.timestamp(),
+                name=name,
+                ready=False,
+                removed=False,
+                reserved=reserved,
+                type=upload_type,
+                user=user,
+            )
+
+            session.add(upload)
+
+            await session.flush()
+
+            upload.name_on_disk = f"{upload.id}-{upload.name}"
+
+            upload = upload.to_dict()
+
+            await session.commit()
+
+        return Upload(**upload)
```

Comment:
You are using the same class here where you need both the SQLAlchemy and core models.

In the subtraction data layer I am importing the `SubtractionFile` SQL model like this:
```py
from virtool.subtractions.models import SubtractionFile as SQLSubtractionFile
```
This allows us to differentiate between the two models. You can do the same for uploads.